nifi-nar-bundles/nifi-airtable-bundle/nifi-airtable-processors/src/main/java/org/apache/nifi/processors/airtable/QueryAirtableTable.java

```diff
@@ -0,0 +1,422 @@
+        airtableRestService = new AirtableRestService(webClientServiceProvider.getWebClientService(), apiUrl, apiKey, baseId, tableId);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final JsonRecordReaderFactory schemaRecordReaderFactory = context.getProperty(SCHEMA_READER).asControllerService(JsonRecordReaderFactory.class);
+        final RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+        final Integer maxRecordsPerFlowFile = context.getProperty(MAX_RECORDS_PER_FLOW_FILE).evaluateAttributeExpressions().asInteger();
+        final Integer outputBatchSize = context.getProperty(OUTPUT_BATCH_SIZE).evaluateAttributeExpressions().asInteger();
+
+        final StateMap state;
+        try {
+            state = context.getStateManager().getState(Scope.CLUSTER);
+        } catch (IOException e) {
+            throw new ProcessException("Failed to get cluster state", e);
+        }
+
+        final String lastRecordFetchDateTime = state.get(LAST_RECORD_FETCH_TIME);
+        final String currentRecordFetchDateTime = OffsetDateTime.now().minusSeconds(QUERY_LAG_SECONDS).format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
+
+        final AirtableGetRecordsParameters getRecordsParameters = buildGetRecordsParameters(context, lastRecordFetchDateTime, currentRecordFetchDateTime);
+        final byte[] recordsJson;
+        try {
+            recordsJson = airtableRestService.getRecords(getRecordsParameters);
+        } catch (RateLimitExceededException e) {
+            context.yield();
+            throw new ProcessException("REST API rate limit exceeded while fetching initial Airtable record set", e);
+        }
+
+        final FlowFile flowFile = session.create();
+        final Map<String, String> originalAttributes = flowFile.getAttributes();
+
+        final RecordSchema recordSchema;
+        final RecordSchema writerSchema;
+        try {
+            final ByteArrayInputStream recordsStream = new ByteArrayInputStream(recordsJson);
+            recordSchema = schemaRecordReaderFactory.createRecordReader(flowFile, recordsStream, getLogger()).getSchema();
+            writerSchema = writerFactory.getSchema(originalAttributes, recordSchema);
+        } catch (MalformedRecordException | IOException | SchemaNotFoundException e) {
+            throw new ProcessException("Couldn't get record schema", e);
+        }
+
+        session.remove(flowFile);
+
+        final List<FlowFile> flowFiles = new ArrayList<>();
+        int totalRecordCount = 0;
+        final AirtableJsonTreeRowRecordReaderFactory recordReaderFactory = new AirtableJsonTreeRowRecordReaderFactory(getLogger(), recordSchema);
+        try (final AirtableRecordSet airtableRecordSet = new AirtableRecordSet(recordsJson, recordReaderFactory, airtableRestService, getRecordsParameters)) {
```

Comment:
It looks like this approach needs to be re-evaluated. Reading the JSON byte array to determine the schema, and then reading it again to process the records is not ideal for memory or performance. If it is not possible to infer the schema after parsing the first record using existing components, then it would be worth evaluating a new implementation that is capable of performing both actions without requiring multiple reads. This approach should also allow reading the JSON as a stream, instead of a byte array.